### PR TITLE
Analytical derivatives in discrete pdf/hfuncs 

### PR DESCRIPTION
**Pull request by tnagler**
Main change: difference quotients coming up in pdf/hfunc computations is replaced by analytical derivatives when evaluation points are close.

This serves two purposes:
* better numerical stability when a discrete "cell" has tiny probability; this comes up frequently in higher vine trees over discrete variables
* automatic handling of continuous distributions with atoms, like zero inflated distributions.

### CODE REVIEW

File: cmake/compilerDefOpt.cmake [hunk 1]
```diff
@@ -38,7 +38,7 @@ if(NOT WIN32)
         set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-const-variable ")
         set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fcolor-diagnostics")
         if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 14.0)
-            set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wc++17-attribute-extensions")
+            set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wc++17-attribute-extensions -Wno-array-bounds")
```
Comment (tvatter) [inline cmake/compilerDefOpt.cmake:41]: Why ?

**Comment by tvatter** [comment]
Is this one ready btw ?

**Comment by tnagler** [comment]
Yes it's ready.

File: include/vinecopulib/bicop/implementation/abstract.ipp [hunk 2]
```diff
@@ -157,15 +157,19 @@ AbstractBicop::pdf_c_d(const Eigen::MatrixXd& u)
...
-    ;
-    pdf = (hfunc2_raw(umax) - hfunc2_raw(umin)).cwiseQuotient(udiff);
   } else {
     udiff = (u.col(1) - u.col(3)).cwiseAbs();
-    pdf = (hfunc1_raw(umax) - hfunc1_raw(umin)).cwiseQuotient(udiff);
   }
 
   for (Eigen::Index i = 0; i < u.rows(); i++) {
-    if (udiff(i) < 1e-3) {
+    if (udiff(i) > 5e-3) {
```
Comment (tvatter) [inline include/vinecopulib/bicop/implementation/abstract.ipp:165]: Is this one a typo ?

**Comment by tvatter** [comment]
Oh, I had missed 10a8e26290b881e85d7c60bc2ef6e5a7d5380c98 ! LGTM then

**Pull request merged**